Changing from Scenarios to Daml Script in all the learning scenarios

nemanja-da opened the pull request:
- Changed the tutorials to use `Daml Script` instead of `scenarios` for testing
- Preview can be found [here](https://beta.katacoda.com/nemanja). Scenarios that were changed are:
   - Fundamentals -> Propose/Accept Pattern
   - Fundamentals -> Contract Keys
   - Fundamentals -> Template authorization 
   - Fundamentals -> Data types and imports
 - Please double-check the changes to make sure that the above mentioned ones are really all that were changed
 - Still need to change all the images (will do that next)

nemanja-da committed:
Changing from Scenarios to Daml Script in all the learning scenarios

robin-da commented on fundamental-concepts/contract-keys/4_using_contract_keys.md:
```diff
@@ -2,60 +2,88 @@ All Daml update actions that work on contract IDs also have a counter part for c
 ![exercise_by_key](assets/exercise_by_key.png)
 
-We can also check this by fetching the contract with `fetchByKey`:
+We can also check this by fetching the contract with `fetchByKey` in Daml. For illustrative purpose we will use `queryContractKey` to show how this done in Daml Script. Before appending the code please delete the `return()` statement.
```
`how this _is_ done`

robin-da commented on fundamental-concepts/contract-keys/4_using_contract_keys.md:
```diff
@@ -2,60 +2,88 @@ All Daml update actions that work on contract IDs also have a counter part for c
-Remove the last line and append
+- `queryContractKey` takes as first argument the `template` name (`@User`), the `Party` that will make the query as the second argument (`alice`), and the contract key as the third argument (`(alice)`)
+- `fetchByKey` takes only two argument, i.e., the template name and the contract key
+- Both actions return a tuple, where the first factor is the actual contract ID and the second factor is the
```
`queryContractKey` returns an `Optional ` not a tuple.

nemanja-da replied:
Changed it to 

> - `queryContractKey` returns an `Optional` tuple (consisting of the contract ID and the contract data)

robin-da commented on fundamental-concepts/contract-keys/4_using_contract_keys.md:
```diff
@@ -2,60 +2,88 @@ All Daml update actions that work on contract IDs also have a counter part for c
+- `lookupByKey` takes the same two arguments as `fetchByKey`, i.e., the template name and the contract key
+- `lookupByKey` is similar to `fetchByKey`, but instead of failing the transaction, it will return a
+`None` if the given key does not exist.
+- `lookupByKey` does not give any information of whether a contract with a given key exists or not.
```
I think this line is wrong. It's `fetchByKey` that does not give any information about existence of a contract.

nemanja-da replied:
I took this from the original Katacoda scenario, you can find it [here on step 4](https://beta.katacoda.com/daml/courses/fundamental-concepts/contract-keys).

> lookupByKey does not give any information of whether a contract with a given key exists or not.

nemanja-da replied:
Just checked the docs, you're right. `fetchByKey` doesn't disclose any info on contract's existence.

robin-da commented on fundamental-concepts/contract-keys/4_using_contract_keys.md:
```diff
@@ -2,60 +2,88 @@ All Daml update actions that work on contract IDs also have a counter part for c
 </pre>
 
-- `exerciseByKey` takes as a first argument the template name of the contract we want to exercise a
+- `exerciseByKeyCmd` maps to `exerciseCmd` in Daml and takes as a first argument the template name of the contract we want to exercise a
```
I don't quite understand this addition. What do you mean by "maps to .. in Daml"?

nemanja-da replied:
Removed that part, should be cleared now

robin-da commented on fundamental-concepts/contract-keys/4_using_contract_keys.md:
```diff
@@ -2,60 +2,88 @@ All Daml update actions that work on contract IDs also have a counter part for c
-`lookupByKey` is similar to `fetchByKey`, but instead of failing the transaction, it will returns a
-`None` if the given key does not exist.
+  let contractId = fst $ fromSome userContractIdAndContract
+  let contract = snd $ fromSome userContractIdAndContract
```
I'd try not to use `fromSome` because it will fail if `userContractidAndContract` is a `None`.

nemanja-da replied:
How would you extract it then?

robin-da reviewed (approved):
There's a couple of corrections I'd make. It would also be nice to give a reasoning for why there is only `queryContractKey` in Daml script.

nemanja-da committed:
Update 4_using_contract_keys.md

Addressing Robin's comments

nemanja-da committed:
Explained why Daml has only query method

Asked Moritz and got the explanation on why there's only query method for both lookup and fetch

nemanja-da commented:
> There's a couple of corrections I'd make. It would also be nice to give a reasoning for why there is only `queryContractKey` in Daml script.

Got this as an explanation from Moritz and added it:

> In Daml lookup and fetch have different authorization rules. Doing a transactional lookup can be quite expensive. Daml Script doesn’t force anyone to validate anything so we can give you the more powerful lookup version where you can handle failures transactionally.

nemanja-da committed:
fetchByKey doesn't return any info on contract's existence

nemanja-da committed:
Added a failure message for extracting the contract key and id
